inbm/dispatcher-agent/tests/ai/snapshot_genai_unit_test.py

```diff
@@ -0,0 +1,2079 @@
+        snapshot = DebianBasedSnapshot(trtl, "update", dispatcher_broker, snap_num, False, reboot_device)
+
+        # Mocking the _rollback_and_delete_snap method
+        snapshot._rollback_and_delete_snap = Mock()
+
+        # Calling the recover method with the test parameters
+        snapshot.recover(rebooter, time_to_wait_before_reboot)
+
+        # Asserting that the _rollback_and_delete_snap method was called the expected number of times
+        if snap_num and snap_num != "0":
+            snapshot._rollback_and_delete_snap.assert_called_once()
+        else:
+            snapshot._rollback_and_delete_snap.assert_not_called()
+
+        # Asserting that the reboot method was called the expected number of times
+        assert rebooter.reboot.call_count == expected_reboot_call_count
+
+#DO NOT DELETE THIS LINE - TestDebianBasedSnapshotRevert
+'''
+ADD HUMAN FEEDBACK BELOW:
+
+'''
+class TestDebianBasedSnapshotRevert:
+
+    @pytest.mark.parametrize('snap_num, time_to_wait_before_reboot, should_reboot', [
+        # Test when snap_num is valid and not "0", should call _rollback_and_delete_snap and reboot
+        ("1", 10, True),
+        # Test when snap_num is "0", should not call _rollback_and_delete_snap and not reboot
+        ("0", 10, False),
+        # Test when snap_num is None, should not call _rollback_and_delete_snap and not reboot
+        (None, 10, False),
+    ])
+    def test_revert(self, snap_num, time_to_wait_before_reboot, should_reboot):
+        # Create a mock for the Rebooter object
+        mock_rebooter = Mock()
+
+        # Create a DebianBasedSnapshot instance with the test snap_num
+        snapshot = DebianBasedSnapshot(Mock(), "update", Mock(), snap_num, False, False)
+
+        # Mock the _rollback_and_delete_snap method
+        snapshot._rollback_and_delete_snap = Mock()
+
+        # Mock the time.sleep function
+        with patch('time.sleep') as mock_sleep:
+            # Call the revert method with the test parameters
+            snapshot.revert(mock_rebooter, time_to_wait_before_reboot)
+
+        # Check if _rollback_and_delete_snap was called
+        if should_reboot:
+            snapshot._rollback_and_delete_snap.assert_called_once()
+        else:
+            snapshot._rollback_and_delete_snap.assert_not_called()
+
+        # Check if the rebooter's reboot method was called
+        if should_reboot:
+            mock_rebooter.reboot.assert_called_once()
+        else:
+            mock_rebooter.reboot.assert_not_called()
+
+        # Check if time.sleep was called with the correct parameter
+        if should_reboot:
+            mock_sleep.assert_called_once_with(time_to_wait_before_reboot)
+        else:
+            mock_sleep.assert_not_called()
+
+#DO NOT DELETE THIS LINE - TestDebianBasedSnapshotRollbackAndDeleteSnap
+'''
+ADD HUMAN FEEDBACK BELOW:
+
+'''
+class TestDebianBasedSnapshotRollbackAndDeleteSnap:
```

Comment:
To go back to Pull Request Summary table, Click on the [link](https://github.com/intel/intel-inb-manageability/pull/489#)

<h3>Class: DebianBasedSnapshot |Function: _rollback_and_delete_snap</h3>

[Source Code Link](https://github.com/intel/intel-inb-manageability/blob/NEX_MAXIMUS_UT/test-specific-unit-test/inbm/dispatcher-agent/dispatcher/sota/snapshot.py?plain=1#L168)

[Unit Test Case Code Link](https://github.com/intel/intel-inb-manageability/blob/NEX_MAXIMUS_UT/test-specific-unit-test/inbm/dispatcher-agent/tests/ai/snapshot_genai_unit_test.py?plain=1#L159)

<h4>Test Case Scenarios:</h4>
The unit test `test_rollback_and_delete_snap` in the class `TestDebianBasedSnapshotRollbackAndDeleteSnap` generates seven test cases based on the parameters provided to the `pytest.mark.parametrize` decorator. Here are the test cases along with their explanations:

1. Test Case: `test_rollback_and_delete_snap[123-sota_rollback_return0-delete_snapshot_return0-expected_telemetry_calls0]`
   - Explanation: This test case checks the scenario where `snap_num` is a valid snapshot number ("123"), the `sota_rollback` method succeeds (returns `(0, None)`), and the `delete_snapshot` method also succeeds. The expected telemetry calls are 'SOTA attempting rollback', 'Rollback succeeded', and 'Snapshot cleanup succeeded'.

2. Test Case: `test_rollback_and_delete_snap[123-sota_rollback_return1-delete_snapshot_return1-expected_telemetry_calls1]`
   - Explanation: This test case checks the scenario where `snap_num` is a valid snapshot number ("123"), but the `sota_rollback` method fails (returns `(1, "Error message")`). The expected telemetry calls are 'SOTA attempting rollback' and 'SOTA rollback failed: Error message'.

3. Test Case: `test_rollback_and_delete_snap[0-sota_rollback_return2-delete_snapshot_return2-expected_telemetry_calls2]`
   - Explanation: This test case checks the scenario where `snap_num` is "0", so the rollback is skipped. The expected telemetry call is 'SOTA rollback skipped'.

4. Test Case: `test_rollback_and_delete_snap[None-sota_rollback_return3-delete_snapshot_return3-expected_telemetry_calls3]`
   - Explanation: This test case checks the scenario where `snap_num` is None, so the rollback is skipped. The expected telemetry call is 'SOTA rollback skipped'.

5. Test Case: `test_rollback_and_delete_snap[abc-sota_rollback_return4-delete_snapshot_return4-expected_telemetry_calls4]`
   - Explanation: This test case checks the scenario where `snap_num` is a non-numeric string ("abc"), so the rollback is skipped. The expected telemetry call is 'SOTA rollback skipped'.

6. Test Case: `test_rollback_and_delete_snap[99999999999999999999-sota_rollback_return5-delete_snapshot_return5-expected_telemetry_calls5]`
   - Explanation: This test case checks the scenario where `snap_num` is a very large number ("99999999999999999999"), the `sota_rollback` method succeeds, and the `delete_snapshot` method also succeeds. The expected telemetry calls are 'SOTA attempting rollback', 'Rollback succeeded', and 'Snapshot cleanup succeeded'.

7. Test Case: `test_rollback_and_delete_snap[-123-sota_rollback_return6-delete_snapshot_return6-expected_telemetry_calls6]`
   - Explanation: This test case checks the scenario where `snap_num` is a negative number ("-123"), so the rollback is skipped. The expected telemetry call is 'SOTA rollback skipped'.

Each test case creates a `DebianBasedSnapshot` instance with the specified `snap_num`, calls the `_rollback_and_delete_snap` method, and checks that the `telemetry` method of the `DispatcherBroker` is called with the expected arguments.

<h4>Short summary about the testcases:</h4>
=========================== short test summary info ============================

FAILED tmp_code_5c9d05aeb03adb941736c59154898c24.py::TestDebianBasedSnapshotRollbackAndDeleteSnap::test_rollback_and_delete_snap[abc-sota_rollback_return4-delete_snapshot_return4-expected_telemetry_calls4] - AssertionError: assert ['SOTA attemp...up succeeded'] == ['SOTA rollback skipped']

  

  At index 0 diff: 'SOTA attempting rollback' != 'SOTA rollback skipped'

  Left contains 2 more items, first extra item: 'Rollback succeeded'

  

  Full diff:

    [

  -     'SOTA rollback skipped',

  +     'SOTA attempting rollback',

  +     'Rollback succeeded',

  +     'Snapshot cleanup succeeded',

    ]

FAILED tmp_code_5c9d05aeb03adb941736c59154898c24.py::TestDebianBasedSnapshotRollbackAndDeleteSnap::test_rollback_and_delete_snap[-123-sota_rollback_return6-delete_snapshot_return6-expected_telemetry_calls6] - AssertionError: assert ['SOTA attemp...up succeeded'] == ['SOTA rollback skipped']

  

  At index 0 diff: 'SOTA attempting rollback' != 'SOTA rollback skipped'

  Left contains 2 more items, first extra item: 'Rollback succeeded'

  

  Full diff:

    [

  -     'SOTA rollback skipped',

  +     'SOTA attempting rollback',

  +     'Rollback succeeded',

  +     'Snapshot cleanup succeeded',

    ]

================== 2 failed, 5 passed, 90 deselected in 0.17s ==================



<h4>Test Execution Results:</h4>

| Total Cases    | 7  |
| -------------- | -- |
| Passed Cases   | 5  |
| Failed Cases   | 2  |



To go back to Pull Request Summary table, Click on the [link](https://github.com/intel/intel-inb-manageability/pull/489#)